File: voice_detection/vad.py
from __future__ import annotations

from dataclasses import dataclass
from math import log10

import numpy as np

from .types import VoiceActivity


EPS = 1e-9


@dataclass
class EnergyVad:
    # Baseline values shared with the C++ implementation.  More permissive
    # values are not necessarily more sensitive once the adaptive noise floor
    # starts following room noise, so tuning must be opt-in and measured.
    threshold_db: float = 8.0
    min_rms_dbfs: float = -52.0
    noise_alpha: float = 0.96
    initial_noise_dbfs: float = -65.0
    calibration_frames: int = 0

    def __post_init__(self) -> None:
        self.noise_floor_dbfs = float(self.initial_noise_dbfs)
        self._calibration_seen = 0

    def process(self, mono: np.ndarray) -> VoiceActivity:
        mono = np.asarray(mono, dtype=np.float32).reshape(-1)
        rms = float(np.sqrt(np.mean(np.square(mono))) + EPS)
        rms_dbfs = 20.0 * log10(max(rms, EPS))
        if self._calibration_seen < self.calibration_frames:
            if self._calibration_seen == 0:
                self.noise_floor_dbfs = rms_dbfs
            else:
                self.noise_floor_dbfs = max(self.noise_floor_dbfs, rms_dbfs)
            self._calibration_seen += 1
            return VoiceActivity(
                active=False,
                probability=0.0,
                rms_dbfs=float(rms_dbfs),
                noise_floor_dbfs=float(self.noise_floor_dbfs),
                snr_db=0.0,
            )
        snr_db = rms_dbfs - self.noise_floor_dbfs
        active = rms_dbfs >= self.min_rms_dbfs and snr_db >= self.threshold_db

        if not active:
            alpha = min(max(self.noise_alpha, 0.0), 0.999)
            self.noise_floor_dbfs = (alpha * self.noise_floor_dbfs) + ((1.0 - alpha) * rms_dbfs)

        probability = sigmoid((snr_db - self.threshold_db) / 3.0)
        if rms_dbfs < self.min_rms_dbfs:
            probability *= 0.4
        return VoiceActivity(
            active=bool(active),
            probability=float(max(0.0, min(1.0, probability))),
            rms_dbfs=float(rms_dbfs),
            noise_floor_dbfs=float(self.noise_floor_dbfs),
            snr_db=float(snr_db),
        )


def sigmoid(value: float) -> float:
    return float(1.0 / (1.0 + np.exp(-value)))

File: src/vad.cpp
#include "voice_detection/vad.hpp"

#include "voice_detection/math_utils.hpp"

#include <cmath>
#include <numeric>

namespace voice_detection
{

EnergyVad::EnergyVad(float threshold_db, float min_rms_dbfs, int calibration_frames)
: threshold_db_(threshold_db), min_rms_dbfs_(min_rms_dbfs),
  calibration_frames_(std::max(0, calibration_frames))
{
}

VoiceActivity EnergyVad::process(const std::vector<float> & mono)
{
  if (mono.empty()) {
    return {};
  }
  double energy = 0.0;
  for (const float sample : mono) {
    energy += static_cast<double>(sample) * static_cast<double>(sample);
  }
  const float rms = static_cast<float>(std::sqrt((energy / static_cast<double>(mono.size())) + 1e-12));
  const float rms_dbfs = 20.0F * std::log10(std::max(rms, 1e-9F));
  if (calibration_seen_ < calibration_frames_) {
    if (calibration_seen_ == 0) {
      noise_floor_dbfs_ = rms_dbfs;
    } else {
      noise_floor_dbfs_ = std::max(noise_floor_dbfs_, rms_dbfs);
    }
    ++calibration_seen_;
    return VoiceActivity{false, 0.0F, rms_dbfs, noise_floor_dbfs_, 0.0F};
  }
  const float snr_db = rms_dbfs - noise_floor_dbfs_;
  const bool active = rms_dbfs >= min_rms_dbfs_ && snr_db >= threshold_db_;
  if (!active) {
    noise_floor_dbfs_ = (noise_alpha_ * noise_floor_dbfs_) + ((1.0F - noise_alpha_) * rms_dbfs);
  }
  float probability = sigmoid((snr_db - threshold_db_) / 3.0F);
  if (rms_dbfs < min_rms_dbfs_) {
    probability *= 0.4F;
  }
  return VoiceActivity{
    active,
    clamp(probability, 0.0F, 1.0F),
    rms_dbfs,
    noise_floor_dbfs_,
    snr_db};
}

}  // namespace voice_detection
